File: p103m/zigzag_level_order.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
const Tree = require('binary_tree');
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var zigzagLevelOrder = function (root) {
  if (root == null) return [];
  let [queue, ans, leftToRight] = [[root], [], true];
  while (queue.length > 0) {
    const [level, len] = [[], queue.length];
    for (let i = 0; i < len; ++i) {
      let node = queue.shift();
      if (leftToRight) level.push(node.val);
      else level.unshift(node.val);
      if (node.left) queue.push(node.left);
      if (node.right) queue.push(node.right);
    }
    leftToRight = !leftToRight;
    ans.push(level);
  }
  return ans;
};
// TEST
[
  '#',
  '1,#,#',
  '1,2,#,#,3,#,#',
  '1,2,3,4,#,#,#,#,#',
  '1,#,2,#,3,#,4,#,#',
  '1,#,2,3,4,#,#,#,#',
  '1,#,2,3,#,#,4,#,#',
  '3,9,#,#,20,15,#,#,7,#,#',
  '3,9,#,11,#,#,20,15,#,#,7,#,#',
].forEach((t) => {
  let tree = Tree.deserialize(t);
  console.log('Zigzag level order of', t, '->', zigzagLevelOrder(tree));
});

File: p103m/zigzag_level_order.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from typing import List
from local_packages.binary_tree import TreeNode


class Solution:
    def zigzagLevelOrder(self, root: TreeNode) -> List[List[int]]:
        if root is None:
            return []
        queue, ans, dir = [root], [], 1
        while queue:
            ans.append([node.val for node in queue[::dir]])
            dir *= -1
            next_level = []
            for node in queue:
                if node.left:
                    next_level.append(node.left)
                if node.right:
                    next_level.append(node.right)
            queue = next_level
        return ans


# TESTS
tests = [
    ("#", []),
    ("1,#,#", [[1]]),
    ("1,2,#,#,3,#,#", [[1], [3, 2]]),
    ("1,2,3,4,#,#,#,#,#", [[1], [2], [3], [4]]),
    ("1,#,2,#,3,#,4,#,#", [[1], [2], [3], [4]]),
    ("1,#,2,3,4,#,#,#,#", [[1], [2], [3], [4]]),
    ("1,#,2,3,#,#,4,#,#", [[1], [2], [3, 4]]),
    ("3,9,#,#,20,15,#,#,7,#,#", [[3], [20, 9], [15, 7]]),
    ("3,9,#,11,#,#,20,15,#,#,7,#,#", [[3], [20, 9], [11, 15, 7]]),
]
for t in tests:
    sol = Solution()
    actual = sol.zigzagLevelOrder(TreeNode.deserialize(t[0]))
    print("Zigzag level order of", t[0], "->", actual)
    assert actual == t[1]
